File: src/main/lib/db/entities/track.ts
import { EntitySchema } from 'typeorm';

import { Track } from '../../../../preload/types/harmony';

export const TrackEntity = new EntitySchema<Track>({
  name: 'track',
  columns: {
    id: {
      type: String,
      primary: true,
    },
    path: {
      type: String,
      nullable: false,
    },
    title: {
      type: String,
      nullable: false,
    },
    artist: {
      type: String,
      nullable: true,
    },
    album: {
      type: String,
      nullable: true,
    },
    genre: {
      type: String,
      nullable: true,
    },
    year: {
      type: Number,
      nullable: true,
    },
    duration: {
      type: Number,
      nullable: false,
    },
    bitrate: {
      type: Number,
      nullable: true,
    },
    comment: {
      type: String,
      nullable: true,
    },
    bpm: {
      type: Number,
      nullable: true,
    },
    initialKey: {
      type: String,
      nullable: true,
    },
    rating: {
      type: 'json',
      nullable: true,
    },
  },
});

File: src-tauri/src/libs/track.rs
// AIDEV-NOTE: Track model and related types
// Matches the TypeScript Track interface in src/preload/types/harmony.ts
// JSON serialization for rating and waveformPeaks stored as TEXT in SQLite

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TrackRating {
  pub source: Option<String>,
  pub rating: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Track {
  pub id: String,
  pub path: String,
  pub title: String,
  pub artist: Option<String>,
  pub album: Option<String>,
  pub genre: Option<String>,
  pub year: Option<i32>,
  pub duration: i64, // milliseconds
  pub bitrate: Option<i32>,
  pub comment: Option<String>,
  pub bpm: Option<i32>,
  pub initial_key: Option<String>,
  pub rating: Option<TrackRating>,
  pub label: Option<String>,
  pub waveform_peaks: Option<Vec<f64>>, // ~300 normalized values (0-1)
  pub added_at: Option<i64>,            // Unix timestamp in milliseconds
  pub url: Option<String>,
}

impl Track {
  /// Generate a deterministic track ID from file path
  /// Uses SHA256 hash of the lowercase absolute path
  pub fn generate_id(path: &str) -> String {
    use sha2::{Digest, Sha256};
    let mut hasher = Sha256::new();
    hasher.update(path.to_lowercase().as_bytes());
    format!("{:x}", hasher.finalize())
  }
}
